content/advent-2016/testing-distributed-systems-in-go.md

```diff
@@ -0,0 +1,185 @@
++++
+author = ["Gyu-Ho Lee"]
+date = "2016-12-27T08:00:00+00:00"
+title = "Testing distributed systems in Go"
+series = ["Advent 2016"]
++++
+
+
+### What is etcd
+
+[etcd](https://github.com/coreos/etcd) is key-value store for the most critical data of distributed systems. Some use cases are: Applications running on [cotainer linux by CoreOS](https://coreos.com/why) gets automatic Linux kernel updates. CoreOS uses etcd to store semaphore values to make sure only subset of cluster are rebooting at any given time. [Kubernetes](http://kubernetes.io/docs/whatisk8s) uses etcd to store cluster states for service discovery and cluster management, and it uses watch API to monitor critical configuration changes. Consistency is the key to ensure that services correctly schedule and operate.
+
+Reliability and robustness is etcd's **highest** priority. This post will explain how etcd is tested under various failure conditions.
+
+
+### Testing "distributed" is difficult
+
+A distributed system connects multiple independent computing nodes, coordinating work through message passing. This definition is deceptive in its simplicity; Leslie Lamport, inventor of Paxos, offered a more realistic definition, *a distributed system is one in which the failure of a computer you didn't even know existed can render your own computer unusable*. Building a stable modern distributed system can be daunting; so many moving parts makes the system inherently complex and ultimately difficult to adequately test.
```

Comment:
`A distributed system connects multiple independent computing nodes, coordinating work through message passing.` =>
'A distributed system consists of multiple independent computing nodes where they coordinate work through passing messages to each other.'